### DISCUSSION TITLE
What exactly is "dirty" in the engine?

### DISCUSSION THREAD

**Discussion by mjarduk** [OpenTTD/OpenTTD#12230] [Q&A] [answered]
Hey uhh I was browsing through the code to see how what works and I've been finding something about "dirty" in gui widgets
Assuming its about redrawing the screen I came here to ask what is the benefit of it over re-rendering the same gui?

**Comment by PeterN** [comment, marked as answer]
Yes, marking windows or widgets dirty lets the window system know that part of the screen needs to be redrawn, because information has changed. In game with developer tools enabled, you can press Ctrl-I to see which viewport areas are being redrawn, although this doesn't show regular widgets.

Without that the game would need to redraw everything on screen every frame, and that is very much an unnecessary performance burden.

So the benefit is improved performance.

**Reply by mjarduk** [reply]
So in a nutshell you store it as a texture and redraw it when its moved or something's been clicked right?

**Reply by michicc** [reply]
If you call the screen a texture, then yes :) OpenTTD is an old-school game and it uses persistent frame content.

**Reply by mjarduk** [reply]
well i saw some mallocs and interpreted it as every window is stored as an array of pixels 

other than that my question has been answered
